app/merginapi.cpp

```diff
@@ -64,13 +240,213 @@ ProjectList MerginApi::parseProjectsData(const QByteArray &data)
+            QJsonValue meta = projectMap.value("meta");
+            p.name = projectMap.value("name").toString();
+            if (meta.isObject()) {
+                QJsonObject metaObject = meta.toObject();
+                int size = metaObject.value("size").toInt(0);
+                int filesCount = metaObject.value("files_count").toInt(0);
+                p.status = getProjectStatus(mDataDir + p.name, size, filesCount);
+            }
+
+            QString created = projectMap.value("created").toString();
             p.info = QDateTime::fromString(created, Qt::ISODateWithMs).toString();
             result << std::make_shared<MerginProject>(p);
         }
     }
     return result;
 }
+
+bool MerginApi::cacheProjectsData(const QByteArray &data)
+{
+    QFile file(mDataDir + "projectsCache.txt");
+    createPathIfNotExists(mDataDir + "projectsCache.txt");
+    if (!file.open(QIODevice::WriteOnly)) {
+        return false;
+    }
+
+    QTextStream stream(&file);
+    stream << data;
+    file.close();
+
+    return true;
+}
+
+void MerginApi::handleDataStream(QNetworkReply* r, QString projectDir)
+{
+    // Read content type from reply's header
+    QByteArray contentType;
+    QString contentTypeString;
+    QList<QByteArray> headerList = r->rawHeaderList();
+    // QByteArray.compare was introduced in Qt 5.12; String conversion needed for Android build
+    QString headString;
+    foreach(QByteArray head, headerList) {
+        headString = QString::fromStdString(head.toStdString());
+        if (headString.compare("Content-Type") == 0) {
+            contentType = r->rawHeader(head);
+            contentTypeString = QString::fromStdString(contentType.toStdString());
+        }
+    }
+
+    // Read boundary hash from content types
+    QString boundary;
+    QRegularExpression re;
+    re.setPattern("[^;]+; boundary=(?<boundary>.+)");
+    QRegularExpressionMatch match = re.match(contentTypeString);
+    if (match.hasMatch()) {
+        boundary = match.captured("boundary");
+    }
+
+    // plus some safety trashold
+    int boundarySize = boundary.length() + 8;
+    QRegularExpression boundaryPattern("(\r\n)?--" + boundary + "\r\n");
+    QRegularExpression headerPattern("Content-Disposition: form-data; name=\"(?P<name>[^'\"]+)\"(; filename=\"(?P<filename>[^\"]+)\")?\r\n(Content-Type: (?P<content_type>.+))?\r\n");
+    QRegularExpression endPattern("(\r\n)?--" + boundary +  "--\r\n");
+
+    QByteArray data;
+    QString dataString;
+    QString activeFilePath;
+
+    while (true) {
+        QByteArray chunk = r->read(CHUNK_SIZE);
+        if (chunk.isEmpty()) {
+            // End of stream - write rest of data to active file
+            if (!activeFilePath.isEmpty()) {
+                QRegularExpressionMatch endMatch = endPattern.match(data);
+                int tillIndex = data.indexOf(endMatch.captured(0));
+                saveFile(data.left(tillIndex), activeFilePath, true);
```

Comment:
This way of iteratively opening file, appending some data and closing it again is not very efficient - with larger files it may be slowing down things a lot. I would suggest to keep the file open, write the chunks and then close it once for all.